HHS/Head-Start-TTADP
# [TTAHUB-3614] Grant replacement race condition bugfix

## Description of change

### The problem

During the debugging for https://jira.acf.gov/browse/TTAHUB-3614, a race condition was discovered that results in duplicate entries in `GrantReplacements` and `GrantReplacementTypes`. I don't believe this resolves 3614, but this fix should go in now.

In this PR we remove the possibility of a race condition and introduce a migration to remove the duplicates that have been created.

The migration:

- The duplicate `GrantReplacementTypes` are removed in favor of the first one created.
- When a `GrantReplacementTypes` entry is removed, any `GrantReplacements` that reference it are updated to use the id of the remaining `GrantReplacementTypes` entry.
- Any duplicate rows in `GrantReplacements` are removed, keeping only the first one created. Non-duplicates are distinct on `replacedGrantId`, `replacingGrantId`, **and** `grantReplacementTypeId`.

## How to test

- Verify the updated logic looks good. The only thing that really changed there is replacing a Promise.all in favor of a for..of loop to ensure sequential processing.
- Download a recent prod dump and run the migration.


There should only be 11 GrantReplacementTypes after migration:

```csv
id   name                             createdAt           updatedAt           deletedAt mapsTo 
---- -------------------------------- ------------------- ------------------- --------- ------ 
1    Relinquished                     2024-10-28 11:48:10 2024-10-28 11:48:10 (null)    (null) 
2    Terminated                       2024-10-28 11:48:10 2024-10-28 11:48:10 (null)    (null) 
3    Unknown                          2024-10-28 11:48:10 2024-10-28 11:48:10 (null)    (null) 
4    Replaced                         2024-10-28 11:48:10 2024-10-28 11:48:10 (null)    (null) 
5    CDI Competitive Replacement      2024-10-29 04:01:32 2024-10-29 04:01:32 (null)    (null) 
6    CDI                              2024-10-29 04:01:32 2024-10-29 04:01:32 (null)    (null) 
94   DRS Non-Competitive Continuation 2024-10-29 04:01:32 2024-10-29 04:01:32 (null)    (null) 
108  BF Conversion                    2024-10-29 04:01:32 2024-10-29 04:01:32 (null)    (null) 
109  Successor in Interest            2024-10-29 04:01:32 2024-10-29 04:01:32 (null)    (null) 
118  DRS Competitive Continuation     2024-10-29 04:01:32 2024-10-29 04:01:32 (null)    (null) 
1126 Consolidated                     2024-10-29 04:01:33 2024-10-29 04:01:33 (null)    (null) 
```

## Issue(s)

* https://jira.acf.gov/browse/TTAHUB-3614


## Checklists

### Every PR



### Before merge to main


### Production Deploy


### After merge/deploy
